do not print mangled name for C functions

plicease opened the pull request:
Consider the C++ code:

```c++
extern "C" int foo(int a[5]) { return 0; }
int bar(int a[5]) { return 0; }
``` 

The output for these two functions is currently:

```xml
  <Function id="_7" name="foo" returns="_14" context="_1" location="f1:1" file="f1" line="1" mangled="_Z3foo">
    <Argument name="a" type="_15" original_type="_16" location="f1:1" file="f1" line="1"/>
  </Function>
  <Function id="_8" name="bar" returns="_14" context="_1" location="f1:2" file="f1" line="2" mangled="_Z3barPi">
    <Argument name="a" type="_15" original_type="_16" location="f1:2" file="f1" line="2"/>
  </Function>
```

The `foo` function includes a nonsensical mangled name even though it is a C function.

```
% objdump -t simple.o  

simple.o:     file format elf64-x86-64

SYMBOL TABLE:
0000000000000000 l    df *ABS*	0000000000000000 simple.C
0000000000000000 l    d  .text	0000000000000000 .text
0000000000000010 g     F .text	000000000000000c _Z3barPi
0000000000000000 g     F .text	000000000000000c foo


```

This patch skips the mangled name for non C++ functions.  Alternatives I could be for C functions to instead:

 1. set the mangled name to `""`, although this conflicts with C++ functions that have `__float128`
 2. set the manged name to be the same as `name`.

With this patch the XML instead looks like:

```xml
  <Function id="_7" name="foo" returns="_14" context="_1" location="f1:1" file="f1" line="1">
    <Argument name="a" type="_15" original_type="_16" location="f1:1" file="f1" line="1"/>
  </Function>
  <Function id="_8" name="bar" returns="_14" context="_1" location="f1:2" file="f1" line="2" mangled="_Z3barPi">
    <Argument name="a" type="_15" original_type="_16" location="f1:2" file="f1" line="2"/>
  </Function>
```

I apologize for not including any tests.  If this approach is broadly accessible I would be happy to try and include one.  Thank you for your consideration.

bradking commented:
@plicease this looks good.  Please update the test suite accordingly.

First, let's add a C-language test.  Try copying `test/input/Function.cxx` to `test/input/Function.c` and updating `test/CMakeLists.txt` to add a new test using it with `castxml_test_output_c(Function)`.  Then run the test suite with `TEST_UPDATE=1` in the environment to generate the expected test output (it should fail the new test the first time and work after that).

Then, also add the test for C linkage in a C++ source.  Copy `test/input/Function.cxx` to `test/input/Function-C.cxx` and follow a similar procedure to add the test.

plicease commented:
@bradking thanks for the feedback and the pointers  I've updated the existing tests (which are affected by the change).  I think this part looks correct.  I've also added simple c linkage tests for functions in C++ and C source.  I used different names, because calling the C test Function with a input of Function.C seemed to conflict with the C++ test by the same name.  I'm happy to change the name again if what I have chosen doesn't fit with the naming scheme.

It also occurred to me variables should also get this fix, since global variables within a namespace get mangled as well.  I can add that to this PR.

bradking commented:
@plicease good work.  Please go ahead and add variables to this PR too.

plicease commented:
@bradking updated for variables.  I think what I am doing is correct, that is if the variable is in the top level of the translation unit it shouldn't be mangled (for either C++ or C).  Using the C++ linkage (`CXXDestructorDecl`) seemed to be the wrong thing from experimentation.

The existing tests seem reasonable for C++, and In think just needed updating (I eyeballed all the Variable* tests, including the ones that didn't change).  I added a C test for good measure, although it is largely identical to the C++ test.  Let me know if there are any other test cases that should be added.

bradking commented on src/Output.cxx:
```diff
@@ -1970,7 +1971,12 @@ void ASTVisitor::OutputVarDecl(clang::VarDecl const* d, DumpNode const* dn)
+  if (!isTranslationUnit)
+  {
+    this->PrintMangledAttribute(d);
+  }
```
Shouldn't this still depend on the ABI (C or C++) rather than on the namespace?  IIRC one can use `extern "C"` inside a namespace and it ignores the namespace for symbol name purposes.

plicease replied:
Ah good point.  Looking at just the `CXXLanguageLinkage` broke tests that shouldn't have changed.  At least the way I tried first.  I've added a failing test so I can see if I can address this test case.

plicease replied:
I'm stumped on how to do this actually.  The reason I am checking the namespace rather than the linkage is because as far as I can tell global variables are only mangled in a namespace.  That is:

```c++
extern "C" int c_linkage;   // C linkage, not mangled
int cpp_linkage;            // C++ linkage, not manged
namespae ns {
extern "C" int c_linkage2;  // C linkage, not manged
int cpp_linkage;            // C++ linkage manged
}
```

I verified each of these using `g++` and `nm`.

My original implementation takes these cases into account except for the last one.  I thought that explictly checking for `extern "C"` might work using `->isExternC()`:


```diff
--- a/src/Output.cxx
+++ b/src/Output.cxx
@@ -2085,8 +2085,7 @@ void ASTVisitor::OutputVarDecl(clang::VarDecl const* d, DumpNode const* dn)
   }
 
   bool const isTranslationUnit = clang::isa<clang::TranslationUnitDecl>(d->getDeclContext());
-  if (!isTranslationUnit)
-  {
+  if (!isTranslationUnit && !d->isExternC()) {
     this->PrintMangledAttribute(d);
   }
   this->PrintAttributesAttribute(d);
```

but it still fails the test in 14da45accf02d528f0e1cbe2d6a0e6954aac134c which is this code:

```c++
namespace ns {
extern "C" int start;
}
```

I've verified using `g++` and `nm` that `start` is not mangled (as you would expect), but I would also expect that the function `isExternC()` would refer to the `extern "C"` decl.  Is libclang wrong?  Or perhaps I misunderstand its intent.

`->isExternC()` or `d->getLanguageLinkage() == clang::CXXLanguageLinkage` only do the "right" (?) thing for variables if the file is C.

bradking commented:
Closing in favor of #200.
